### DISCUSSION TITLE
what is the proper way to stop pocketbase?

### DISCUSSION THREAD

**Discussion by bit-lang** [pocketbase/pocketbase#6086] [Q&A]
Currently I'm using Ctrl-c to interrupt, but feeling uneasy to do so.
Is there a suggested method to stop pocketbase, e.g. when running as a go module, use app.Stop() in go code?

**Comment by ganigeorgiev** [comment]
Ctrl-c should work because on most systems this will send `SIGINT` signal and we have graceful shutdown handling for it in https://github.com/pocketbase/pocketbase/blob/9efd68ff4cb3e3dd39cd4f57c8ad747aafa3484a/pocketbase.go#L170 (`os.Interrupt` is an alias for `syscall.SIGINT`).

If for some reason you want to terminate the PocketBase process programmatically from within the application itself you can try triggering the [`OnTerminate` hook](https://pocketbase.io/docs/go-event-hooks/#onterminate), something like:

```go
event := new(core.TerminateEvent)
event.App = app
err := app.OnTerminate().Trigger(event, func(e *core.TerminateEvent) error {
    // this is optional as it will force closing the db connections
    // aka. after the terminate event if you want to be able to access the app DB(), Settings(), etc.
    // you'll have to call again app.Start() or app.Bootstrap()
    return e.App.ResetBootstrapState()
})
```
But the above very rarely would have to be called by the end users and I'm not sure if it is worth adding a `Stop` helper just for that (or at least I don't see the use case for it).

In all cases even if the application is abruptly stopped technically it still should be fine as long as you don't manually delete or mix the sidecar db WAL files (`.db-wal` and `.db-shm`).

**Reply by bit-lang** [reply]
thank you for the example.
sorry if my understanding is wrong, but isn't the above app.OnTerminate().Trigger() only happens when user already presses Ctrl-c?

my intention is to be able to terminate the PocketBase process programmatically, using something maybe like app.Stop() or app.Shutdown(), similar to golang http server's func server.Shutdown().

is that possible?

**Reply by ganigeorgiev** [reply]
Please read my comment above after the _"If for some reason you want to terminate the PocketBase process programmatically from within the application ..."_.

In any case, without more context why what you are trying to do is even needed in the first place, I don't know how to help further.

Feel free to elaborate but for now I'm not planning investing time on this.

**Reply by bit-lang** [reply]
> Please read my comment above after the _"If for some reason you want to terminate the PocketBase process programmatically from within the application ..."_.
> 
> In any case, without more context why what you are trying to do is even needed in the first place, I don't know how to help further.
> 
> Feel free to elaborate but for now I'm not planning investing time on this.

one such context is when using PocketBase as a module, one could decide when to start and stop PocketBase, while keeping other app logics running.

**Reply by ganigeorgiev** [reply]
@bit-lang Either there is a language barrier that prevents us to understand each other or you are not reading my comment (_and your last reply doesn't really provide much more information from the description in your original question_)_.

In any case, as mentioned in my initial reply, there is no Stop/Shutdown helper method but you can easily create your own and call the following code inside it:

```go
event := new(core.TerminateEvent)
event.App = app
err := app.OnTerminate().Trigger(event, func(e *core.TerminateEvent) error {
    // this is optional as it will force closing the db connections
    // aka. after the terminate event if you want to be able to access the app DB(), Settings(), etc.
    // you'll have to call again app.Start() or app.Bootstrap()
    return e.App.ResetBootstrapState()
})
```

Other than this I really don't know how to help further.

**Reply by nikhiljohn10** [reply]
I am trying to build an app where I can start and stop the server like Caddy does. I tried the following code to stop the pb server and exit gracefully. But it is stuck in terminal. Once this stop endpoint is called, the terminal does not even accept Ctrl+C to exit the terminal. Any help would be appreciated.

```go
func AddRoutes(app *pocketbase.PocketBase) {
	app.OnServe().BindFunc(func(se *core.ServeEvent) error {

		se.Router.GET("/stop", func(e *core.RequestEvent) error {
			event := new(core.TerminateEvent)
			event.App = app
			return app.OnTerminate().Trigger(event, func(e *core.TerminateEvent) error {
				return e.App.ResetBootstrapState()
			})
		})

		return se.Next()
	})
}

```

**Reply by nikhiljohn10** [reply]
This code works. But will this create any further problem internally? 

```go
func AddStopRoute(app *pocketbase.PocketBase) {
	app.OnServe().BindFunc(func(se *core.ServeEvent) error {

		se.Router.GET("/stop", func(e *core.RequestEvent) error {
			event := new(core.TerminateEvent)
			event.App = app
			return app.OnTerminate().Trigger(event, func(e *core.TerminateEvent) error {
				err := e.App.ResetBootstrapState()
				if err == nil {
					syscall.Exit(0)
				}
				return nil
			})
		})

		return se.Next()
	})
}
```
